### PR TITLE
refactor: migrate to next.js and socket.io

### PR DESCRIPTION
**Pull request by drewradcliff**

### CODE REVIEW

File: tsconfig.json
Comment (lewxdev) [inline tsconfig.json:1]: I think tsconfig is easier to maintain if we use [tsconfig/bases](https://github.com/tsconfig/bases), but this can be a separate PR

```shell
npm i -D @tsconfig/{next,strictest}
```

```jsonc
// path: tsconfig.json
{
  "extends": ["@tsconfig/next", "@tsconfig/strictest"],
  "compilerOptions": {
    "paths": {
      "@/*": ["./*"]
    }
  }
}

File: tailwind.config.ts [hunk 1]
```diff
@@ -1,2 +1,20 @@
...
+  ],
+  theme: {
+    extend: {
+      backgroundImage: {
+        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
+        "gradient-conic":
+          "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))",
+      },
+    },
+  },
```
Comment (lewxdev) [inline tailwind.config.ts]: this is a leftover from the shitty next boilerplate :tm:
```suggestion
  theme: {
    extend: {},
  },
```

File: tailwind.config.ts [hunk 2]
```diff
@@ -1,2 +1,20 @@
-// this file is just here for intellisense
-// see: https://github.com/tw-in-js/vscode-twind-intellisense/issues/24#issuecomment-1670633807
+import type { Config } from "tailwindcss";
+
+const config: Config = {
+  content: [
+    "./pages/**/*.{js,ts,jsx,tsx,mdx}",
+    "./components/**/*.{js,ts,jsx,tsx,mdx}",
+    "./app/**/*.{js,ts,jsx,tsx,mdx}",
+  ],
```
Comment (lewxdev) [inline tailwind.config.ts:8]: not in the tailwind docs, but if this uses standard bracket expansion, this should be like this (needs to be tested):
```suggestion
  content: [
    "./{app,components,pages}/**/*.{js,ts,jsx,tsx,mdx}",
  ],
```

if using `src` only, we can just use `"./src/**/*.{js,ts,jsx,tsx,mdx}"`
see: https://tailwindcss.com/docs/guides/nextjs

File: src/game.ts
Comment (lewxdev) [inline src/game.ts:1]: we probably want to restore this

File: src/client.tsx
Comment (lewxdev) [inline src/client.tsx:1]: we could migrate this over to next.js land

File: src/api.ts
Comment (lewxdev) [inline src/api.ts:1]: we can reuse this in api routes (or, more likely, just let the socket.io handle this)

File: public/vercel.svg
Comment (lewxdev) [inline public/vercel.svg:1]: this should be deleted

File: public/next.svg
Comment (lewxdev) [inline public/next.svg:1]: this should be deleted

File: postcss.config.mjs
Comment (lewxdev) [inline postcss.config.mjs:1]: god I can't wait for tailwind v4 🥲

File: package.json [hunk 3]
```diff
@@ -1,49 +1,29 @@
...
+  },
+  "dependencies": {
+    "next": "14.2.5",
+    "react": "^18",
+    "react-dom": "^18",
+    "socket.io": "^4.7.5",
+    "socket.io-client": "^4.7.5"
+  },
+  "devDependencies": {
+    "@flydotio/dockerfile": "^0.5.8",
```
Comment (lewxdev) [inline package.json:26]: genuinely not sure if [the dockerfile generator](https://www.npmjs.com/package/@flydotio/dockerfile) is necessary here (especially now that we have a working one) but this may be okay to keep to try and make the modification process smoother for that

File: package.json [hunk 4]
```diff
@@ -1,49 +1,29 @@
...
+    "@types/node": "^20",
+    "@types/react": "^18",
+    "@types/react-dom": "^18",
+    "eslint": "^8",
+    "eslint-config-next": "14.2.5",
+    "postcss": "^8",
+    "tailwindcss": "^3.4.1",
+    "typescript": "^5"
+  }
```
Comment (lewxdev) [inline package.json]: we can make the following changes here to support the code we had previously as well (with some improvements). see:
- [Next.js | Configuring: ESLint | Usage With Other Tools | Prettier](https://nextjs.org/docs/app/building-your-application/configuring/eslint#prettier)
- [Next.js | Configuring: ESLint | Usage With Other Tools | lint-staged](https://nextjs.org/docs/app/building-your-application/configuring/eslint#lint-staged)
- [Prettier | Configuration File | EditorConfig](https://prettier.io/docs/en/configuration#editorconfig)

```jsonc
// path: package.json

{
  "name": "mmmines",
  "private": true,
  "scripts": {
    "dev": "node app/server.mjs",
    "build": "next build",
    "start": "NODE_ENV=production node app/server.mjs",
    "lint": "next lint",
    "format": "prettier --cache --write .",
    "prepare": "husky"
  },
  "dependencies": {
    "@upstash/redis": "^1.32.0",
    "lodash": "^4.17.21",
    "next": "14.2.5",
    "react": "^18",
    "react-dom": "^18",
    "socket.io": "^4.7.5",
    "socket.io-client": "^4.7.5"
  },
  "devDependencies": {
    "@commitlint/cli": "^19.3.0",
    "@commitlint/config-conventional": "^19.2.2",
    "@flydotio/dockerfile": "^0.5.8",
    "@types/lodash": "^4.17.6",
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "eslint": "^8",
    "eslint-config-next": "14.2.5",
    "eslint-config-prettier": "^9.1.0",
    "husky": "^9.0.11",
    "lint-staged": "^15.2.7",
    "postcss": "^8",
    "prettier": "^3.3.2",
    "tailwindcss": "^3.4.1",
    "typescript": "^5"
  },
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional"
    ]
  },
  "eslintConfig": {
    "extends": [
      "next/core-web-vitals",
      "prettier"
    ]
  }
}
```
```js
// path: lint-staged.config.mjs

import path from "node:path";

export default {
  "*": "prettier --cache --ignore-unknown --write",
  "*.{js,jsx,ts,tsx}": (filenames) =>
    `next lint --fix ${filenames
      .map((filename) => `--file ${path.relative(process.cwd(), filename)}`)
      .join(" ")}`,
};
```
```ini
# path: .editorconfig
# learn more: https://editorconfig.org

root = true

[*]
charset = utf-8
end_of_line = lf
indent_size = 2
indent_style = space
insert_final_newline = true
trim_trailing_whitespace = true
```

File: next.config.mjs
Comment (lewxdev) [inline next.config.mjs:1]: do we even need this config if it's empty? 🤔

File: fly.toml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# fly.toml app configuration file generated for mmmines-v4 on 2024-07-14T21:24:15-04:00
+#
+# See https://fly.io/docs/reference/configuration/ for information about how to use this file.
+#
+
+app = 'mmmines-v4'
+primary_region = 'dfw'
```
Comment (lewxdev) [inline fly.toml]:
```suggestion
# see: https://fly.io/docs/reference/configuration/

app = 'mmmines'
primary_region = 'ord'
```
also, according to that docs link, it seems like we can use alternative formats. `yaml` would be cleaner, but `json` would be preferred if there's a json schema out there that provides intellisense

File: .eslintrc.json
Comment (lewxdev) [inline .eslintrc.json:1]: at the moment, this should be in the `package.json`, such tiny file, much wasted space

File: .gitignore [hunk 6]
```diff
@@ -1,48 +1,36 @@
...
+
+# local env files
+.env*.local
+
+# vercel
+.vercel
+
+# typescript
+*.tsbuildinfo
+next-env.d.ts
```
Comment (lewxdev) [inline .gitignore]:
```suggestion
# based on the node gitignore file from github/gitignore
# see: https://github.com/github/gitignore/blob/main/Node.gitignore
# learn more: https://git-scm.com/docs/gitignore

# build
build

# caches
.cache
.eslintcache
.node_repl_history
.npm

# dependencies
node_modules

# environment
.env
.env*.local
.dev.vars

# logs
logs
*.log
*-debug.log*
*-error.log*

# runtime
pids
*.pid
*.seed
*.pid.lock

# miscellaneous
.temp
.DS_Store
*.tgz

# addons
# see: https://nodejs.org/api/addons.html
build/Release

# reports
# see: https://nodejs.org/api/report.html
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json

# next.js
.next
out

# typescript
*.tsbuildinfo
next-env.d.ts
```

File: .husky/commit-msg
Comment (lewxdev) [inline .husky/commit-msg:1]: this should be restored

File: .husky/pre-commit
Comment (lewxdev) [inline .husky/pre-commit:1]: this should be restored

File: .vscode/extensions.json
Comment (lewxdev) [inline .vscode/extensions.json:1]: this should be restored

File: .vscode/settings.json [hunk 7]
```diff
@@ -1,10 +0,0 @@
-{
-	"editor.defaultFormatter": "esbenp.prettier-vscode",
-	"editor.formatOnSave": true,
-	"files.insertFinalNewline": true,
-	"files.trimFinalNewlines": true,
-	"files.trimTrailingWhitespace": true,
-	"files.associations": {
-		".dev.vars*": "properties"
-	}
-}
```
Comment (lewxdev) [inline .vscode/settings.json]: this should be restored, but you can make the following changes (if we also add an `.editorconfig`)
```jsonc
// path: .vscode/settings.json

{
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": "explicit"
  },
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "editor.formatOnSave": true,
}
```

File: app/favicon.ico
Comment (lewxdev) [inline app/favicon.ico:1]: this can be deleted

File: app/globals.css
Comment (lewxdev) [inline app/globals.css:1]: re: comment about Tailwind v4 🥲

File: app/layout.tsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import type { Metadata } from "next";
+import { Inter } from "next/font/google";
```
Comment (lewxdev) [inline app/layout.tsx]: since we're using next, it should be really easy to drop in [geist](https://vercel.com/font#get)

File: app/layout.tsx [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+export default function RootLayout({
+  children,
+}: Readonly<{
+  children: React.ReactNode;
+}>) {
+  return (
+    <html lang="en">
+      <body className={inter.className}>{children}</body>
```
Comment (lewxdev) [inline app/layout.tsx]: ew this boilerplate looks ugly, try this?
```suggestion
export default function RootLayout(props: Readonly<React.PropsWithChildren>) {
  return (
    <html lang="en">
      <body className={inter.className}>{props.children}</body>
```

File: app/layout.tsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import type { Metadata } from "next";
+import { Inter } from "next/font/google";
+import "./globals.css";
+
+const inter = Inter({ subsets: ["latin"] });
+
+export const metadata: Metadata = {
+  title: "Create Next App",
+  description: "Generated by create next app",
```
Comment (lewxdev) [inline app/layout.tsx]:
```suggestion
export const metadata: Metadata = {
  title: "mmmines",
```

File: app/page.tsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    socket.on("connect", onConnect);
+    socket.on("disconnect", onDisconnect);
+    socket.on("message", onMessage);
+
+    return () => {
+      socket.off("connect", onConnect);
+      socket.off("disconnect", onDisconnect);
+      socket.off("message", onMessage);
+    };
+  }, []);
```
Comment (lewxdev) [inline app/page.tsx]: I would extract the socket.io stuff into it's own hook. also see [Socket.IO | Typescript](https://socket.io/docs/v4/typescript/) for properly defining the types we expect between the frontend and server code.

File: README.md
Comment (lewxdev) [inline README.md:1]: honestly, we don't need this boilerplate readme, I'd just use something as simple as the following

```md
<!-- path: README.md -->

# mmmines
```

File: app/server.mjs
Comment (lewxdev) [inline app/server.mjs:1]: if you're willing to try bun again, now that I think about it, we should've been [using `Bun.serve` to run the http server](https://bun.sh/guides/http/hot), not `"node:http"`. maybe one or both of us can look into this 🤷🏽

File: .dockerignore
Comment (lewxdev) [inline .dockerignore:1]: this will need to be updated to the new `.gitignore` if that's updated. I don't like the idea of this extra file we have to "maintain". at worst, we can symlink `.dockerignore` -> `.gitignore`. at best, we can configure the ignore behavior for docker to use `.gitignore`

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+# Set production environment
+ENV NODE_ENV="production"
+
+
+# Throw-away build stage to reduce size of final image
+FROM base as build
+
+# Install packages needed to build node modules
+RUN apt-get update -qq && \
+    apt-get install --no-install-recommends -y build-essential node-gyp pkg-config python-is-python3
```
Comment (lewxdev) [inline Dockerfile:21]: I wonder if the bun issue we were running into had to do with different dependencies necessary here 🤔

**Review by lewxdev — CHANGES_REQUESTED** [review]

File: app/server.mjs
Comment (lewxdev) [reply app/server.mjs:1]: feel free to disregard this, it's simply not ready yet.
see: https://github.com/socketio/socket.io/issues/4969

File: tailwind.config.ts (hunk 2 above)
Comment (drewradcliff) [reply tailwind.config.ts:8]: If you want to make an issue for this or we can discuss later if we want to move to src directory. For now, the default is fine.

File: Dockerfile (hunk 12 above)
Comment (lewxdev) [reply Dockerfile:21]: done with bun, no more nonsense

File: README.md [hunk 13]
```diff
@@ -1,8 +1,36 @@
...
+- [Next.js Documentation](https://nextjs.org/docs) - learn about Next.js features and API.
+- [Learn Next.js](https://nextjs.org/learn) - an interactive Next.js tutorial.
+
+You can check out [the Next.js GitHub repository](https://github.com/vercel/next.js/) - your feedback and contributions are welcome!
+
+## Deploy on Vercel
+
+The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/new?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
+
+Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.
```
Comment (lewxdev) [inline README.md]:
```suggestion
# mmmines
```

File: .dockerignore
Comment (drewradcliff) [reply .dockerignore:1]: Symlink is probably overkill. We could also add some CI/CD steps to copy `.gitignore` -> `.dockerignore` before deployment. But we can talk about a solution in the future.

File: .eslintrc.json
Comment (drewradcliff) [reply .eslintrc.json:1]: It seems this will not be available in v9 so we might as well keep the config file
https://github.com/eslint/eslint/discussions/18131

File: .editorconfig
Comment (drewradcliff) [inline .editorconfig:1]: maybe we should look into using this with prettier and if there's any conflicts

Comment (lewxdev) [reply .editorconfig:1]: this should be fine, editorconfig is preferred since it has features prettier doesn't support by default, like `insert_final_newline`. I've also not experienced any conflicts in my years of using this.

> If a [`.editorconfig` file](https://editorconfig.org/) is in your project, Prettier will parse it and convert its properties to the corresponding Prettier configuration. This configuration will be overridden by `.prettierrc`, etc.

see: https://prettier.io/docs/en/configuration#editorconfig

File: .dockerignore
Comment (lewxdev) [reply .dockerignore:1]: I don't think symlink is overkill since it can be tracked in the repo. set once and forget, no need to deal with docker config or CI/CD (that seems like more overkill to me)

File: .eslintrc.json
Comment (lewxdev) [reply .eslintrc.json:1]: I didn't know this, but this should be fine since we won't be upgrading (`next lint` is not compatible with v9, hence the starter still using `eslint@^8`). v9 makes a LOT of changes that are generally not compatible with anything out there right now, namely [flat config](https://eslint.org/blog/2022/08/new-config-system-part-2/). I'd like to keep the configuration files to a minimum if possible

File: .gitignore [hunk 14]
```diff
@@ -46,3 +35,20 @@ pids
...
+# see: https://nodejs.org/api/report.html
+report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
+
+# next.js
+.next
+out
+
+# typescript
+*.tsbuildinfo
+next-env.d.ts
```
Comment (lewxdev) [inline .gitignore]: oops, needs final newline 😵‍💫

**Comment by lewxdev** [comment]
if we ever want to deploy to a (p)review environment, we can:
see: https://fly.io/docs/blueprints/review-apps-guide/

**Review by lewxdev — APPROVED** [review]

**Comment by lewxdev** [comment]
im gonna merge it
